File: sdk/android/tracelet-sdk/src/main/kotlin/com/ikolvi/tracelet/sdk/StateManager.kt
package com.ikolvi.tracelet.sdk

import android.content.Context
import android.content.SharedPreferences
import com.ikolvi.tracelet.sdk.model.TrackingMode

/**
 * Tracks the runtime state of the Tracelet plugin.
 *
 * State includes: enabled, trackingMode, schedulerEnabled, odometer,
 * didLaunchInBackground, didDeviceReboot.
 */
class StateManager(context: Context) {

    companion object {
        private const val PREFS_NAME = "com.tracelet.state"
        private const val KEY_ENABLED = "enabled"
        private const val KEY_TRACKING_MODE = "trackingMode"
        private const val KEY_SCHEDULER_ENABLED = "schedulerEnabled"
        private const val KEY_ODOMETER = "odometer"
        private const val KEY_IS_MOVING = "isMoving"
        private const val KEY_DID_LAUNCH_IN_BACKGROUND = "didLaunchInBackground"
        private const val KEY_DID_DEVICE_REBOOT = "didDeviceReboot"
        private const val KEY_LAST_LOCATION_TIME = "lastLocationTime"
        private const val KEY_LAST_PERIODIC_LAT = "lastPeriodicLat"
        private const val KEY_LAST_PERIODIC_LNG = "lastPeriodicLng"
        private const val KEY_SPEED_MOTION_STATE = "speedMotionState"
        private const val KEY_SPEED_LOW_COUNT = "speedLowCount"
        private const val KEY_SPEED_WAKE_COUNT = "speedWakeCount"
        private const val KEY_SPEED_LAST_TRANSITION = "speedLastTransition"
    }

    private val prefs: SharedPreferences =
        context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)

    var enabled: Boolean
        get() = prefs.getBoolean(KEY_ENABLED, false)
        set(value) = prefs.edit().putBoolean(KEY_ENABLED, value).apply()

    /** The current tracking mode — see [TrackingMode]. */
    var trackingMode: TrackingMode
        get() = TrackingMode.fromInt(prefs.getInt(KEY_TRACKING_MODE, 0))
        set(value) = prefs.edit().putInt(KEY_TRACKING_MODE, value.value).apply()

    var schedulerEnabled: Boolean
        get() = prefs.getBoolean(KEY_SCHEDULER_ENABLED, false)
        set(value) = prefs.edit().putBoolean(KEY_SCHEDULER_ENABLED, value).apply()

    var odometer: Double
        get() = java.lang.Double.longBitsToDouble(
            prefs.getLong(KEY_ODOMETER, java.lang.Double.doubleToLongBits(0.0))
        )
        set(value) = prefs.edit().putLong(
            KEY_ODOMETER, java.lang.Double.doubleToLongBits(value)
        ).apply()

    var isMoving: Boolean
        get() = prefs.getBoolean(KEY_IS_MOVING, false)
        set(value) = prefs.edit().putBoolean(KEY_IS_MOVING, value).apply()

    var didLaunchInBackground: Boolean
        get() = prefs.getBoolean(KEY_DID_LAUNCH_IN_BACKGROUND, false)
        set(value) = prefs.edit().putBoolean(KEY_DID_LAUNCH_IN_BACKGROUND, value).apply()

    var didDeviceReboot: Boolean
        get() = prefs.getBoolean(KEY_DID_DEVICE_REBOOT, false)
        set(value) = prefs.edit().putBoolean(KEY_DID_DEVICE_REBOOT, value).apply()

    var lastLocationTime: Long
        get() = prefs.getLong(KEY_LAST_LOCATION_TIME, 0L)
        set(value) = prefs.edit().putLong(KEY_LAST_LOCATION_TIME, value).apply()

    /** Last periodic fix latitude (for odometer computation across worker runs). */
    var lastPeriodicLatitude: Double
        get() = java.lang.Double.longBitsToDouble(
            prefs.getLong(KEY_LAST_PERIODIC_LAT, java.lang.Double.doubleToLongBits(Double.NaN))
        )
        set(value) = prefs.edit().putLong(
            KEY_LAST_PERIODIC_LAT, java.lang.Double.doubleToLongBits(value)
        ).apply()

    // ---------------------------------------------------------------------------
    // Speed-based motion detection state
    // ---------------------------------------------------------------------------

    /** Current speed motion state: "moving", "slowing", or "stationary". */
    var speedMotionState: String?
        get() = prefs.getString(KEY_SPEED_MOTION_STATE, null)
        set(value) = prefs.edit().putString(KEY_SPEED_MOTION_STATE, value).apply()

    /** Consecutive low-speed fix count (SLOWING state). */
    var speedLowCount: Int
        get() = prefs.getInt(KEY_SPEED_LOW_COUNT, 0)
        set(value) = prefs.edit().putInt(KEY_SPEED_LOW_COUNT, value).apply()

    /** Consecutive wake-speed fix count (STATIONARY state). */
    var speedWakeCount: Int
        get() = prefs.getInt(KEY_SPEED_WAKE_COUNT, 0)
        set(value) = prefs.edit().putInt(KEY_SPEED_WAKE_COUNT, value).apply()

    /** Epoch millis of last speed-motion state transition. */
    var speedLastTransition: Long
        get() = prefs.getLong(KEY_SPEED_LAST_TRANSITION, 0L)
        set(value) = prefs.edit().putLong(KEY_SPEED_LAST_TRANSITION, value).apply()

    /** Last periodic fix longitude (for odometer computation across worker runs). */
    var lastPeriodicLongitude: Double
        get() = java.lang.Double.longBitsToDouble(
            prefs.getLong(KEY_LAST_PERIODIC_LNG, java.lang.Double.doubleToLongBits(Double.NaN))
        )
        set(value) = prefs.edit().putLong(
            KEY_LAST_PERIODIC_LNG, java.lang.Double.doubleToLongBits(value)
        ).apply()

    /** Adds [distance] meters to the cumulative odometer. */
    fun addOdometer(distance: Double) {
        odometer += distance
    }

    /** Returns the full state as a map for Dart consumption. */
    fun toMap(config: Map<String, Any?>? = null): Map<String, Any?> {
        return mapOf(
            "enabled" to enabled,
            "trackingMode" to trackingMode.value,
            "schedulerEnabled" to schedulerEnabled,
            "odometer" to odometer,
            "isMoving" to isMoving,
            "didLaunchInBackground" to didLaunchInBackground,
            "didDeviceReboot" to didDeviceReboot,
            "config" to config,
        )
    }

    /** Resets all state to defaults. */
    fun reset() {
        prefs.edit().clear().apply()
    }
}

File: sdk/ios/Sources/TraceletSDK/StateManager.swift
import Foundation

/// Tracks plugin runtime state: enabled, tracking mode, odometer, motion state, etc.
public final class StateManager {
    private let defaults: UserDefaults
    private let prefix = "com.tracelet.state."

    public var enabled: Bool {
        get { defaults.bool(forKey: prefix + "enabled") }
        set { defaults.set(newValue, forKey: prefix + "enabled") }
    }

    /// The current tracking mode (continuous, geofences, periodic).
    public var trackingMode: TrackingMode {
        get { TrackingMode.fromInt(defaults.integer(forKey: prefix + "trackingMode")) }
        set { defaults.set(newValue.rawValue, forKey: prefix + "trackingMode") }
    }

    public var schedulerEnabled: Bool {
        get { defaults.bool(forKey: prefix + "schedulerEnabled") }
        set { defaults.set(newValue, forKey: prefix + "schedulerEnabled") }
    }

    public var isMoving: Bool {
        get { defaults.bool(forKey: prefix + "isMoving") }
        set { defaults.set(newValue, forKey: prefix + "isMoving") }
    }

    public var odometer: Double {
        get { defaults.double(forKey: prefix + "odometer") }
        set { defaults.set(newValue, forKey: prefix + "odometer") }
    }

    public var didLaunchInBackground: Bool {
        get { defaults.bool(forKey: prefix + "didLaunchInBackground") }
        set { defaults.set(newValue, forKey: prefix + "didLaunchInBackground") }
    }

    public var lastLocationTime: Double {
        get { defaults.double(forKey: prefix + "lastLocationTime") }
        set { defaults.set(newValue, forKey: prefix + "lastLocationTime") }
    }

    /// Last periodic fix latitude (for odometer computation across app restarts).
    /// Returns `NaN` when no periodic fix has been recorded.
    public var lastPeriodicLatitude: Double {
        get {
            guard defaults.object(forKey: prefix + "lastPeriodicLatitude") != nil else {
                return Double.nan
            }
            return defaults.double(forKey: prefix + "lastPeriodicLatitude")
        }
        set {
            if newValue.isNaN {
                defaults.removeObject(forKey: prefix + "lastPeriodicLatitude")
            } else {
                defaults.set(newValue, forKey: prefix + "lastPeriodicLatitude")
            }
        }
    }

    /// Last periodic fix longitude (for odometer computation across app restarts).
    /// Returns `NaN` when no periodic fix has been recorded.
    public var lastPeriodicLongitude: Double {
        get {
            guard defaults.object(forKey: prefix + "lastPeriodicLongitude") != nil else {
                return Double.nan
            }
            return defaults.double(forKey: prefix + "lastPeriodicLongitude")
        }
        set {
            if newValue.isNaN {
                defaults.removeObject(forKey: prefix + "lastPeriodicLongitude")
            } else {
                defaults.set(newValue, forKey: prefix + "lastPeriodicLongitude")
            }
        }
    }

    public var speedMotionState: Int? {
        get {
            let key = prefix + "speedMotionState"
            return defaults.object(forKey: key) as? Int
        }
        set {
            if let val = newValue {
                defaults.set(val, forKey: prefix + "speedMotionState")
            } else {
                defaults.removeObject(forKey: prefix + "speedMotionState")
            }
        }
    }

    public var speedLowCount: Int {
        get { defaults.integer(forKey: prefix + "speedLowCount") }
        set { defaults.set(newValue, forKey: prefix + "speedLowCount") }
    }

    public var speedWakeCount: Int {
        get { defaults.integer(forKey: prefix + "speedWakeCount") }
        set { defaults.set(newValue, forKey: prefix + "speedWakeCount") }
    }

    public var speedLastTransition: Double {
        get { defaults.double(forKey: prefix + "speedLastTransition") }
        set { defaults.set(newValue, forKey: prefix + "speedLastTransition") }
    }

    public init() {
        defaults = UserDefaults.standard
    }

    /// Adds `distance` meters to the cumulative odometer.
    public func addOdometer(distance: Double) {
        odometer += distance
    }

    public func reset() {
        enabled = false
        trackingMode = .continuous
        schedulerEnabled = false
        isMoving = false
        odometer = 0.0
        didLaunchInBackground = false
        lastLocationTime = 0
        lastPeriodicLatitude = .nan
        lastPeriodicLongitude = .nan
    }

    public func toMap(_ config: [String: Any]?) -> [String: Any] {
        return [
            "enabled": enabled,
            "trackingMode": trackingMode.rawValue,
            "schedulerEnabled": schedulerEnabled,
            "isMoving": isMoving,
            "odometer": odometer,
            "didLaunchInBackground": didLaunchInBackground,
            "lastLocationTime": lastLocationTime,
            "config": config as Any,
        ]
    }
}
